src/understand-tidb/transaction.md

```diff
@@ -1 +1,114 @@
+
+```go
+// BatchGet gets all the keys' value from kv-server and returns a map contains key/value pairs.
+// The map will not contain nonexistent keys.
+func (s *tikvSnapshot) BatchGet(ctx context.Context, keys []kv.Key) (map[string][]byte, error) {
+	data, err := s.KVSnapshot.BatchGet(ctx, toTiKVKeys(keys))
+	return data, extractKeyErr(err)
+}
+
+// Get gets the value for key k from snapshot.
+func (s *tikvSnapshot) Get(ctx context.Context, k kv.Key) ([]byte, error) {
+	data, err := s.KVSnapshot.Get(ctx, k)
+	return data, extractKeyErr(err)
+}
+```
+
+For the write SQL statements, they will write data into the transaction memory buffer temporarily until the `commit` operations is triggered. There are 3 main interfaces which will write query data into the memory buffer, they are the [tables](https://github.com/pingcap/tidb/blob/af70762cd52519f025daa5e869ba37465a7fb311/table/table.go#L166) API:
+
+```go
+// Table is used to retrieve and modify rows in table.
+type Table interface {
+	...
+	// AddRecord inserts a row which should contain only public columns
+	AddRecord(ctx sessionctx.Context, r []types.Datum, opts ...AddRecordOption) (recordID kv.Handle, err error)
+
+	// UpdateRecord updates a row which should contain only writable columns.
+	UpdateRecord(ctx context.Context, sctx sessionctx.Context, h kv.Handle, currData, newData []types.Datum, touched []bool) error
+
+	// RemoveRecord removes a row in the table.
+	RemoveRecord(ctx sessionctx.Context, h kv.Handle, r []types.Datum) error
+	...
+}
+```
+
+Every statement will use a `staging buffer` during its execution, if it's successful the staging content will be merged into the transaction memory buffer. For example `AddRecord` will try to write an row into the current statement staging buffer, and the `RemoveRecord` will try to remove an row in the staging statement buffer. The transaction memory buffer will not be affected if the statement has failed.
+
+The memory buffer implementation is wrapped in [memBuffer](https://github.com/pingcap/tidb/blob/af70762cd52519f025daa5e869ba37465a7fb311/store/driver/txn/unionstore_driver.go#L27). The internal implementation is `MemDB` [object](https://github.com/tikv/client-go/blob/4fc565e203a99400d0b080a25a93fb860b3b6fd6/internal/unionstore/memdb.go).
+
+The memory buffer is a ordered map and it provides the `staging` and `discard` utilities. For example the memory content generated by a statement will be discarded if its execution has failed.
+
+
+# The Two Phase Commit
```

Comment:
```suggestion
# The Two-phase Commit
```
I feel like the word "two-phase" is commonly used one. Wikipedia and Spanner's paper both use it.

https://en.wikipedia.org/wiki/Two-phase_commit_protocol